Сдача этапа 2: Бэкенд

borodulex opened the pull request:
Реализовал бэкенд для дипломного проекта:
- Настроил окружение
- Создал схемы и модели ресурсов API
- Создал роуты и контроллеры
- Реализовал аутентификацию и авторизацию
- Настроил логирование
- Описал возвращаемые ошибки централизованным обработчиком
- Поработал над безопасностью
- Задеплоил на сервер

borodulex committed:
feat: preparation for backend development
[+] editorconfig
[+] eslint and config-airbnb-base
[+] gitignore
[+] nodemon hotreload
[+] package.json init and scripts
[+] express
[+] README
[+] init app.js

borodulex committed:
feat: creating schemas and models of API resources
[+] mongoose
[+] validator for schemas
[+] schemas and modules for users, movies

borodulex committed:
feat: config file
[+] add config file for process.env using dotenv
[*] update README

borodulex committed:
feat: centralized error handler
[+] error middleware and error objects

borodulex committed:
feat: creating routes and controllers
[+] user routes and controllers
[+] user _id temp hardcode
[+] body parser

borodulex committed:
feat: movieId int validation

borodulex committed:
feat: creating routes and controllers
[+] movies routes and controllers
[*] movieId validatoin fix

borodulex committed:
refactor: Cast error message

borodulex committed:
fix: error-handler status code

borodulex committed:
feat: creating routes and controllers
[+] routes auth protection using jwt and cookie
[*] config change
[*] README update

borodulex committed:
feat: authentication and authorization
[+] createUser method
[+] http codes move to consts
[*] fix double next() call

borodulex committed:
feat: authentication and authorization
[*] orFail instead of if

borodulex committed:
feat: authentication and authorization
[*] remove auth hardcode
[+] auth functionaluty using jwt and cookie
[+] signin and signout routes
[+] userSchema static method findUserByCredentials

borodulex committed:
feat: non-existent resource routes error handle

borodulex committed:
feat: logger
[+] request and error logger using winston

borodulex committed:
feat: schemas custom comments for errors

borodulex committed:
feat: celebrate validation

borodulex committed:
feat: helmet for Content-Security-Policy

borodulex committed:
feat: rate-limitter

borodulex committed:
feat: README update

borodulex committed:
feat: README update

borodulex committed:
feat: cors middleware

borodulex committed:
feat: connect router to /api path

borodulex committed:
feat: updateBio duplicate email error add

borodulex committed:
feat: README update

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--12.57--> Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [ ] <!--25.45--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [ ] `POST /movies` — создаёт фильм с переданными в теле данными;
  
  Свойство owner не должно валидироваться celebrate. Celebrate валидирует данные, которые передаёт пользователь, свойство owner же добавляется сервером из req.user
  
  - [ ] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  
Удаление необходимо осуществлять по свойству _id фильма. Сейчас происходит попытка удалить фильм по свойству movieId

  - [ ] `POST /signup` — создаёт пользователя с переданными в теле данными;
  
  Нет возможности создать пользователя с именем больше 8 символов.
![image](https://user-images.githubusercontent.com/17438227/128344132-c69451e6-043e-4cd9-8a42-d23b28c48691.png)

  
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--3.64--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--3.64--> Данные валидируются перед добавлением в базу.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--3.64--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--1.5--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--1.5--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--1.5--> Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
- [x] <!--1.5--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--1.5--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--1.5--> Централизованный обработчик ошибок описан в отдельном модуле.
- [ ] <!--1.5--> В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.

Обновление имени пользователя без передачи нового имени заканчивается успешно, хотя должна возвращаться ошибка 400

- [x] <!--1.5--> К серверу можно обратиться по https по адресу домена, указанному в README.md.  
- [ ] <!--1.5--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.83--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.83--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.83--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.83--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.83--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.83--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.

## Количество баллов: 71

borodulex committed:
hotfix: level-1, 1st review
[*] remove movie body celebrate owner validation
[*] deleting movie by _id, instead of movieId
[*] user's name max length - 30
[*] dots in error messages
[*] user bio fields set to required on update validation

borodulex committed:
hotfix: movie _id error text fix

borodulex committed:
hotfix: logger uncomment

borodulex commented:
Исправил замечания из резюме.

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--12.57--> Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [x] <!--25.45--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--3.64--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--3.64--> Данные валидируются перед добавлением в базу.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--3.64--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--1.5--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--1.5--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--1.5--> Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
- [x] <!--1.5--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--1.5--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--1.5--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--1.5--> В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.
- [x] <!--1.5--> К серверу можно обратиться по https по адресу домена, указанному в README.md.  
- [x] <!--1.5--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.83--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.83--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.83--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.83--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.83--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.83--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.

## Количество баллов: 99

- Ошибка CastError может возникнуть, если пользователь передал неправильный id в запрос. 
Данный запрос не получает id от пользователя, то есть данный тип ошибки не сможет возникнуть в запросе. Поэтому проверку на имя err.name === 'CastError' можно убрать из условия.
![image](https://user-images.githubusercontent.com/17438227/128627366-20af0437-388f-4c45-b421-25d90d5d6659.png)
